subsys/usb/host/usbh_shell.c

```diff
@@ -0,0 +1,627 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/usb/usbh.h>
+
+#include "usbh_internal.h"
+#include "usbh_ch9.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbh_shell, CONFIG_USBH_LOG_LEVEL);
+
+#define FOOBAZ_VREQ_OUT		0x5b
+#define FOOBAZ_VREQ_IN		0x5c
+
+USBH_CONTROLLER_DEFINE(uhs_ctx, DEVICE_DT_GET(DT_NODELABEL(zephyr_uhc0)));
+
+const static struct shell *ctx_shell;
+
+static void print_dev_desc(const struct shell *sh,
+			   const struct usb_device_descriptor *const desc)
+{
+	shell_print(sh, "bLength\t\t\t%u", desc->bLength);
+	shell_print(sh, "bDescriptorType\t\t%u", desc->bDescriptorType);
+	shell_print(sh, "bcdUSB\t\t\t%x", desc->bcdUSB);
+	shell_print(sh, "bDeviceClass\t\t%u", desc->bDeviceClass);
+	shell_print(sh, "bDeviceSubClass\t\t%u", desc->bDeviceSubClass);
+	shell_print(sh, "bDeviceProtocol\t\t%u", desc->bDeviceProtocol);
+	shell_print(sh, "bMaxPacketSize0\t\t%u", desc->bMaxPacketSize0);
+	shell_print(sh, "idVendor\t\t%x", desc->idVendor);
+	shell_print(sh, "idProduct\t\t%x", desc->idProduct);
+	shell_print(sh, "bcdDevice\t\t%x", desc->bcdDevice);
+	shell_print(sh, "iManufacturer\t\t%u", desc->iManufacturer);
+	shell_print(sh, "iProduct\t\t%u", desc->iProduct);
+	shell_print(sh, "iSerial\t\t\t%u", desc->iSerialNumber);
+	shell_print(sh, "bNumConfigurations\t%u", desc->bNumConfigurations);
+}
+
+static void print_cfg_desc(const struct shell *sh,
+			   const struct usb_cfg_descriptor *const desc)
+{
+	shell_print(sh, "bLength\t\t\t%u", desc->bLength);
+	shell_print(sh, "bDescriptorType\t\t%u", desc->bDescriptorType);
+	shell_print(sh, "wTotalLength\t\t%x", desc->wTotalLength);
+	shell_print(sh, "bNumInterfaces\t\t%u", desc->bNumInterfaces);
+	shell_print(sh, "bConfigurationValue\t%u", desc->bConfigurationValue);
+	shell_print(sh, "iConfiguration\t\t%u", desc->iConfiguration);
+	shell_print(sh, "bmAttributes\t\t%02x", desc->bmAttributes);
+	shell_print(sh, "bMaxPower\t\t%u mA", desc->bMaxPower * 2);
+}
+
+static void print_desc(const struct shell *sh, const struct net_buf *const buf)
+{
+	struct usb_desc_header *head = (void *)buf->data;
```

Comment:
This function doesn't sanitize received data length. It shouldn't really cause any issues other than simply printing uninitialized data if device responds with outright incorrect descriptor (too short to be descriptor of given type).

Reply:
Added net_buf length checks.